tools/llvm-pta-eval.cpp

```diff
@@ -0,0 +1,697 @@
+  printf("target1=");
+  printName(ptr1.target, 0);
+  printf("\n");
+  printf("target2=");
+  printName(ptr2.target, 0);
+  printf("\n");
+  if (ptr1.target->isUnknownMemory()) return MayAlias;
+  if (ptr2.target->isUnknownMemory()) return MayAlias;
+  if (ptr1.offset == UNKNOWN_OFFSET) return MayAlias;
+  if (ptr2.offset == UNKNOWN_OFFSET) return MayAlias;
+  if (ptr1.target == ptr2.target
+      && ptr1.offset == ptr2.offset) return MustAlias;
+  return NoAlias;
+}
+
+static int check_pointer(const Pointer& ptr, char *name)
+{
+  printf("target %s=", name);
+  printName(ptr.target, 0);
+  if (ptr.target->isUnknownMemory()) {
+    printf("Unknown Ptr\n");
+    return MayAlias;
+  }
+  if (ptr.target->isNull()) {
+    printf("Null Ptr\n");
+    return MayAlias;
+  }
+  printf("\n");
+  return NoAlias;
+}
+static int dump_pointer(const Pointer& ptr, char *name)
+{
+  printf("target %s=", name);
+  printName(ptr.target, 0);
+  printf("\n");
+  return 0;
+}
+static AliasResult doAlias(LLVMPointerAnalysis *pta, 
+			   llvm::Value *V1, llvm::Value*V2)
+{
+  PSNode *p1 = pta->getPointsTo(V1);
+  PSNode *p2 = pta->getPointsTo(V2);
+  int count1 = 0;
+  int count2 = 0;
+  for (const Pointer& ptr1 : p1->pointsTo) {
+    count1++;
+    if (!ptr1.isValid())
+      continue;
+  }
+  for (const Pointer& ptr2 : p2->pointsTo) {
+    count2++;
+    if (!ptr2.isValid())
+      continue;
+  }
+
+  printf("counts = %d %d\n", count1, count2);
+  if (count1 > 1 || count2 > 1) {
+    for (const Pointer& ptr1 : p1->pointsTo) {
+      if (!ptr1.isValid())
+        continue;
+      dump_pointer(ptr1, "1");
+    }
+    for (const Pointer& ptr2 : p2->pointsTo) {
+      if (!ptr2.isValid())
+        continue;
+      dump_pointer(ptr2, "2");
+    }
+    return MayAlias;
+  }
+  Pointer ptr1(UNKNOWN_MEMORY, UNKNOWN_OFFSET);
+  Pointer ptr2(UNKNOWN_MEMORY, UNKNOWN_OFFSET);
+  if (count1 == 0 && count2 == 0) {
+    return NoAlias;
+  }
+  if (count1 == 1) {
+    auto itr1 = p1->pointsTo.begin();
+    ptr1 = *itr1;
+    for (;itr1 !=p1->pointsTo.end();++itr1) {
+      ptr1 = *itr1;
+      if (!ptr1.isValid())
+	continue;
+      break;
+    }
+  }
+  if (count2 == 1) {
+    auto itr2 = p2->pointsTo.begin();
+    ptr2 = *itr2;
+    for (;itr2 !=p2->pointsTo.end();++itr2) {
+      ptr2 = *itr2;
+      if (!ptr2.isValid())
+	continue;
+      break;
+    }
+  }
+  if (count1 == 0) {
+    return check_pointer(ptr2, "2");
+  }
+  if (count2 == 0) {
+    return check_pointer(ptr1, "1");
+  }
+  return compare_pointer(ptr1, ptr2);
+}
+
+
+static llvm::StringRef NOALIAS("NOALIAS"), 
+  MAYALIAS("MAYALIAS"), 
+  MUSTALIAS("MUSTALIAS"),
+  PARTIALALIAS("PARTIALALIAS"),
+  EXPECTEDFAIL_MAYALIAS = ("EXPECTEDFAIL_MAYALIAS"),
+  EXPECTEDFAIL_NOALIAS("EXPECTEDFAIL_NOALIAS");
+
+
+static int test_checkfunc(const llvm::StringRef &fun)
+{
+  if (fun.equals(NOALIAS)) {
+    return 1;
+  } else if (fun.equals(MAYALIAS)) {
+    return 1;
+  } else if (fun.equals(MUSTALIAS)) {
+    return 1;
```

Comment:
Maybe this would be better (only one return 1 statement)
```C
 if (fun.equals(...) ||
     fun.equals(...) ||
     ...)
    return 1;
```

and would it be also possible to use `bool` instead of  `int` as a return value in this case?